parlai/utils/conversations.py

```diff
@@ -93,6 +95,74 @@ def save_metadata(
             f.write(json.dumps(metadata))
 
 
+class Turn(AttrDict):
+    """
+    Utility class for a dialog turn.
+    """
+
+    def __init__(self, id=None, text=None, **kwargs):
+        super().__init__(self, id=id, text=text, **kwargs)
+
+
+class Conversation:
+    """
+    Utility class for iterating through a single episode.
+
+    Used in the context of the Conversations class.
+    """
+
+    def __init__(self, episode):
+        self.episode = episode
+        self.context = episode.get('context')
+        self.metadata_path = episode.get('metadata_path')
+        self.turns = self._build_turns(episode)
+        self.iterator_idx = 0
+
+    def _build_turns(self, episode):
+        turns = []
+        for act_pair in episode['dialog']:
+            for act in act_pair:
+                turns.append(Turn(**act))
+        return turns
+
+    def read(self):
+        print(BAR)
+        high_level = [k for k in self.episode.keys() if k != 'dialog']
+        if high_level:
+            for key in high_level:
+                print(f'{key}: {self.episode[key]}')
+            print(SMALL_BAR)
+
+        for turn in self.turns:
+            print(f'{turn.id}: {turn.text}')
+
+        print(BAR)
+
+    @property
+    def num_turns(self):
+        return len(self.turns)
+
+    def __getitem__(self, index):
+        return self.turns[index]
+
+    def __next__(self):
+        """
+        Return the next conversation.
+        """
+        if self.iterator_idx >= len(self.turns):
+            print('You reached the end of the conversation.')
+            self.reset()  # return the iterator idx to 0
+            return None
```

Comment:
Normally you'd do `raise StopIteration` to indicate we're at the end. Python uses this for for loops...